### PR TITLE
GH-43266: [C#] Add LargeBinary, LargeString and LargeList array types

### PR DESCRIPTION
**Pull request by adamreeve**
### Rationale for this change

See #43266. Note that LargeBinary and LargeString are still limited to 2 GiB buffers, and LargeList is limited to offsets that can be represented as int32.

### What changes are included in this PR?

* Add new Array subtypes: LargeBinaryArray, LargeStringArray and LargeListArray
* Support round-tripping these array types via the IPC format
* Support round-tripping these array types via the C Data Interface
* Improve error messages when importing arrays that are too large via IPC or C Data Interface
* Enable integration tests for the new types 
* Update documentation

### Are these changes tested?

Yes, I've added some basic tests specifically for the new array types, and added these to the test data generator so they're covered by the existing tests for round tripping using IPC and C Data Interface.

### Are there any user-facing changes?

Yes, this is a new user facing feature.

### Implementation notes

* I haven't added builders for these new array types. Given they're added to help with interoperability with other libraries, I wouldn't expect .NET users to build arrays of these types as they provide no other benefit over the non-large types until we have proper large memory support. But I'm happy to add this if it would be useful.
* The new array types share a lot of logic with the non-large types. I considered trying to consolidate this logic by adding a new `BinaryArrayBase<TOffset>` class for example, but I think this would require generic math support to work nicely, and would still complicate the code quite a bit and add extra virtual method call overhead. So I think it's fine to keep these new Array subtypes independent from the non-large types.
* I haven't included support for materializing a LargeStringArray (see #41048). I'm not sure whether there would be a use for this, but it could be added later if needed.
* GitHub Issue: #43266

### CODE REVIEW

File: csharp/src/Apache.Arrow/Arrays/LargeBinaryArray.cs [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+    public override void Accept(IArrowArrayVisitor visitor) => Accept(this, visitor);
+
+    public ArrowBuffer ValueOffsetsBuffer => Data.Buffers[1];
+
+    public ArrowBuffer ValueBuffer => Data.Buffers[2];
+
+    public ReadOnlySpan<long> ValueOffsets => ValueOffsetsBuffer.Span.CastTo<long>().Slice(Offset, Length + 1);
+
+    public ReadOnlySpan<byte> Values => ValueBuffer.Span.CastTo<byte>();
```
Comment (adamreeve) [inline csharp/src/Apache.Arrow/Arrays/LargeBinaryArray.cs:57]: I'm not sure if we want to consider naming this something different like `SmallValues` to help with backwards compatibility when we eventually have something like a `LargeReadOnlySpan<T>` type.

This problem isn't specific to these new array types though. For `PrimitiveArray` for example we'll probably also want to introduce a new "Large" version of the `Values` ReadOnlySpan, so for consistency I think it's fine to keep calling this `Values`. Then we can later add something named like `LargeValues` to all applicable array types.

File: csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs [hunk 2]
```diff
@@ -132,7 +132,13 @@ protected ReadResult ReadMessage()
 
                 Flatbuf.Message message = Flatbuf.Message.GetRootAsMessage(CreateByteBuffer(messageBuff));
 
-                int bodyLength = checked((int)message.BodyLength);
+                if (message.BodyLength > int.MaxValue)
```
Comment (adamreeve) [inline csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs:135]: I don't think there's a need for more specific checks when importing individual arrays as this should cover all scenarios.

I also tested reading data from Flight but that actually fails on the C++ server side with an error that > 2 GB record batches are not supported (see [FlightPayload::Validate](https://github.com/apache/arrow/blob/21238a7f5ebb224410d84ecf32dbe3556cab363a/cpp/src/arrow/flight/types.cc#L207)).

**Review by raulcd — COMMENTED** [review]
hey @adamreeve ! this is great but Isn't the main point of using Large List and Large Binary / Large String to be able to have [offsets represented with int64 ](https://github.com/apache/arrow/blob/main/docs/source/format/Columnar.rst?plain=1#L552)and/or data buffers > 2GiB?
This will allow some compatibility with other languages but on the cases supported we should probably not use Large Binary/String/List and just use "normal" Binary/String/List.

I am missing to see the rationale of the change without supporting the main "feature" of these Large formats but probably I am just missing some context :)

**Comment by CurtHagenlocher** [comment]
@raulcd The primary motivation is integration with Polars, which apparently doesn't support the non-Large versions.

**Comment by adamreeve** [comment]
Yeah as Curt says this is about compatibility with other libraries, but otherwise these don't currently provide any other benefit if you're only using .NET, and you'll get errors if you try to import anything that's too large.

I'm happy to revert the documentation changes if you think they're misleading? Or maybe update the wording to add that this is only to help with interoperability?

**Comment by raulcd** [comment]
> Or maybe update the wording to add that this is only to help with interoperability?

Thanks, that makes sense with the context. I see the value of supporting it now. I think we could try to improve the wording as suggested as it might be misleading for users who might potentially have issues with interoperability when using large formats from other systems/languages if those don't fit the current limitations.

**Comment by adamreeve** [comment]
OK thanks, I've updated the status documentation so it isn't so misleading now

File: csharp/src/Apache.Arrow/Arrays/LargeBinaryArray.cs (hunk 1 above)
Comment (CurtHagenlocher) [reply csharp/src/Apache.Arrow/Arrays/LargeBinaryArray.cs:57]: I have to admit that I like the consistency of having the same members on both classes, but I also wonder at the value (ha ha) of exposing this at all. Someone who needs to get at the underlying buffer can already access ValueBuffer, and this span doesn't have any clear uses.

**Review by CurtHagenlocher — APPROVED** [review]

File: csharp/src/Apache.Arrow/Arrays/LargeBinaryArray.cs (hunk 1 above)
Comment (adamreeve) [reply csharp/src/Apache.Arrow/Arrays/LargeBinaryArray.cs:57]: Good point, not adding this member solves the backwards compatibility problem nicely, and I don't see a great need for it either. I'll remove this.

**Review by CurtHagenlocher — APPROVED** [review]
Thanks!

**Pull request merged**